### PR TITLE
Variable MPI buffer size and cap communication chunk size

### PR DESCRIPTION
**Pull request by josephleekl**
**Context:**
Currently the size of the MPI buffer when using Lightning-Kokkos with MPI is hard coded (each buffer is half the size of the statevector, and 2 buffers are allocated, one for send and one for receive). For large statevector, this consumes lots of memory, and the total sizes needed to send could exceed the MPI size limit (max int).

**Description of the Change:**
Users can set the ratio for the buffer size, where 1 means the total buffer size is equal to the statevector size, and any integer N > 1 denotes the `total buffer size = size of SV // N`. The default for N is 1.

We also chunk transfers limited by the buffer size or the MPI int size count

**Benefits:**
No overflow for MPI. Reduce memory footprint for LK-MPI.

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-121282]

### CODE REVIEW

File: mpitests/test_device.py [hunk 1]
```diff
@@ -132,3 +132,29 @@ def test_unsupported_gate():
     ):
         dev.execute(tape)
         comm.Barrier()
+
+
+@pytest.mark.skipif(device_name != "lightning.kokkos", reason="Only for Lightning-Kokkos")
+def test_comm_buffer_ratio_matches_default():
+    """comm_buffer_ratio is accepted with mpi=True and yields the same result
+    as the default ratio (it changes only how transfers are chunked)."""
```
Comment (josephleekl) [inline mpitests/test_device.py]:
```suggestion
    as the default ratio."""
```

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 2]
```diff
@@ -908,31 +958,32 @@ class StateVectorKokkosMPI final
...
         auto recvbuf_view = (*recvbuf_);
         auto sv_view = (*sv_).getView();
 
-        // Since the buffer is half the size of the state vector, we need to
-        // do two copies
-        for (std::size_t i = 0; i < 2; i++) {
-            std::size_t offset = i * send_size;
+        // Transfer the full local sub-state-vector in chunks no larger than
+        // the comm buffer; with a half-size buffer this is two chunks.
+        for (std::size_t off = 0; off < total_size; off += chunk_size) {
```
Comment (josephleekl) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]:
```suggestion
        for (std::size_t offset = 0; offset < total_size; offset += chunk_size) {
```

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 3]
```diff
@@ -908,31 +958,32 @@ class StateVectorKokkosMPI final
...
+            const std::size_t csize = std::min(chunk_size, total_size - off);
```
Comment (josephleekl) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]:
```suggestion
            const std::size_t csize = std::min(chunk_size, total_size - offset);
```

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 4]
```diff
@@ -908,31 +958,32 @@ class StateVectorKokkosMPI final
...
             // COPY to buffer
             Kokkos::parallel_for(
-                "copy_sendbuf", RangePolicy<KokkosExecSpace>(0, send_size),
+                "copy_sendbuf", RangePolicy<KokkosExecSpace>(0, csize),
                 KOKKOS_LAMBDA(std::size_t buffer_index) {
-                    sendbuf_view(buffer_index) = sv_view(buffer_index + offset);
+                    sendbuf_view(buffer_index) = sv_view(buffer_index + off);
```
Comment (josephleekl) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]:
```suggestion
                    sendbuf_view(buffer_index) = sv_view(buffer_index + offset);
```

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 5]
```diff
@@ -908,31 +958,32 @@ class StateVectorKokkosMPI final
...
-            sendrecvBuffers(dest_mpi_rank, dest_mpi_rank, send_size, 0);
+            sendrecvBuffers(dest_mpi_rank, dest_mpi_rank, csize, 0);
             // COPY FROM BUFFER
-
             Kokkos::parallel_for(
-                "copy_recvbuf", RangePolicy<KokkosExecSpace>(0, send_size),
+                "copy_recvbuf", RangePolicy<KokkosExecSpace>(0, csize),
                 KOKKOS_LAMBDA(std::size_t buffer_index) {
-                    sv_view(buffer_index + offset) = recvbuf_view(buffer_index);
+                    sv_view(buffer_index + off) = recvbuf_view(buffer_index);
```
Comment (josephleekl) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]:
```suggestion
                    sv_view(buffer_index + offset) = recvbuf_view(buffer_index);
```

File: mpitests/test_device.py [hunk 6]
```diff
@@ -132,3 +132,29 @@ def test_unsupported_gate():
     ):
         dev.execute(tape)
         comm.Barrier()
+
+
+@pytest.mark.skipif(device_name != "lightning.kokkos", reason="Only for Lightning-Kokkos")
+def test_comm_buffer_ratio_matches_default():
```
Comment (jzaia18) [inline mpitests/test_device.py]: I think we should parametrize this test and try every ratio from 1-5, to test edge cases where the splitting is either nonexistent or non-uniform

**Review by jzaia18 — APPROVED** [review]
Looks good, just 1 comment about testing

File: mpitests/test_device.py (hunk 6 above)
Comment (josephleekl) [reply mpitests/test_device.py]: updated

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 7]
```diff
@@ -304,13 +310,50 @@ class StateVectorKokkosMPI final
...
     std::size_t getNumGlobalWires() const { return numGlobalQubits_; }
 
     std::size_t getNumLocalWires() const { return numLocalQubits_; }
 
+    /**
+     * @brief Default MPI comm-buffer memory-reduction factor.
+     *
+     * Buffer size = max(1, half_sv / ratio). Two of these buffers (send, recv)
+     * are allocated per process. The default of 1 keeps full half-SV buffers.
+     * larger values trade communication latency for less per-process memory.
```
Comment (josephleekl) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]:
```suggestion
     * Larger values trade communication latency for less per-process memory.
```

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 8]
```diff
@@ -98,6 +98,8 @@ class StateVectorKokkosMPI final
     std::shared_ptr<KokkosVector> sendbuf_;
     MPIManagerKokkos mpi_manager_;
 
+    std::size_t comm_buffer_ratio_ = COMM_BUFFER_RATIO;
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]: suggestion: Maybe `DEFAULT_COMM_BUFFER_RATIO` is a more descriptive naming?

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 9]
```diff
@@ -770,48 +813,60 @@ class StateVectorKokkosMPI final
...
-
-                    sv_view(SV_index) = recvbuf_view(buffer_index);
-                });
-            Kokkos::fence();
+            // Transfer in chunks no larger than the comm buffer; with a
+            // full-size buffer this is a single chunk.
+            for (std::size_t offset = 0; offset < send_size;
+                 offset += chunk_size) {
+                const std::size_t csize =
+                    std::min(chunk_size, send_size - offset);
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp:832]: praise:  Neat handling of boundary.

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 10]
```diff
@@ -770,48 +813,60 @@ class StateVectorKokkosMPI final
             auto sendbuf_view = (*sendbuf_);
             auto recvbuf_view = (*recvbuf_);
             auto sv_view = (*sv_).getView();
-            std::size_t send_size =
+            const std::size_t send_size =
                 exp2((getNumLocalWires() - local_wires_to_swap.size()));
+            // Each MPI_Sendrecv count must fit in a 32-bit int, so clamp the
+            // chunk to the MPI transfer limit.
+            const std::size_t chunk_size = std::min(
+                sendbuf_->extent(0), MPIManagerKokkos::MPI_MAX_TRANSFER_COUNT);
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]: suggestion: Replace `extent(0)` by `size()` to better express intent.

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 11]
```diff
@@ -908,31 +963,36 @@ class StateVectorKokkosMPI final
         std::size_t dest_mpi_rank = getElementIndexInVector(
             mpi_rank_to_global_index_map_target, dest_global_index);
 
-        std::size_t send_size = exp2(getNumLocalWires() - 1);
+        const std::size_t total_size = exp2(getNumLocalWires());
+        // Each MPI_Sendrecv count must fit in a 32-bit int, so clamp the chunk
+        // to the MPI transfer limit.
+        const std::size_t chunk_size = std::min(
+            sendbuf_->extent(0), MPIManagerKokkos::MPI_MAX_TRANSFER_COUNT);
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]: suggestions: `extent(0)` -> `size()`

File: pennylane_lightning/core/simulators/lightning_kokkos/tests/mpi/Test_StateVectorKokkosMPI.cpp [hunk 12]
```diff
@@ -1115,7 +1115,9 @@ TEMPLATE_TEST_CASE("sendrecvBuffers", "[LKMPI]", double, float) {
 
     std::size_t mpi_rank = mpi_manager.getRank();
     std::size_t dest_rank = mpi_rank ^ 1U;
-    std::size_t message_size = 4;
+    // The comm buffer size is set by allocateBuffers (from the ratio), so
+    // transfer exactly what the buffer holds rather than a hard-coded count.
+    std::size_t message_size = sendbuf.extent(0);
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/tests/mpi/Test_StateVectorKokkosMPI.cpp]: suggestion: `extent(0)` -> `size()`

File: pennylane_lightning/core/simulators/lightning_kokkos/utils/MPIManagerKokkos.hpp [hunk 13]
```diff
@@ -107,6 +118,10 @@ class MPIManagerKokkos final : public MPIManager {
                   Kokkos::View<T *> &recvBuf, std::size_t source,
                   std::size_t size, std::size_t tag = 0) {
         MPI_Datatype datatype = getMPIDatatype<T>();
+        PL_ABORT_IF(size > MPI_MAX_TRANSFER_COUNT,
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/utils/MPIManagerKokkos.hpp:121]: question: Does this not apply to the other methods in this file (e.g., `Bcast`)?

File: pennylane_lightning/core/simulators/lightning_kokkos/tests/mpi/Test_StateVectorKokkosMPI.cpp [hunk 14]
```diff
@@ -1130,6 +1132,102 @@ TEMPLATE_TEST_CASE("sendrecvBuffers", "[LKMPI]", double, float) {
...
+    REQUIRE(mpi_manager.getSize() == 4);
+
+    // Default ratio when not specified.
+    SV sv_default(mpi_manager, num_qubits);
+    CHECK(sv_default.getCommBufferRatio() == SV::COMM_BUFFER_RATIO);
+
+    // Custom ratio (kokkos_args default, ratio = 2).
+    SV sv2(mpi_manager, num_qubits, Kokkos::InitializationSettings{}, 2);
+    CHECK(sv2.getCommBufferRatio() == 2);
+    CHECK(sv2.getSendBuffer()->extent(0) ==
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/tests/mpi/Test_StateVectorKokkosMPI.cpp]: suggestion: `extent(0)` -> `size()`

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp [hunk 15]
```diff
@@ -908,31 +963,36 @@ class StateVectorKokkosMPI final
...
+            sendrecvBuffers(dest_mpi_rank, dest_mpi_rank, csize, 0);
             // COPY FROM BUFFER
-
             Kokkos::parallel_for(
-                "copy_recvbuf", RangePolicy<KokkosExecSpace>(0, send_size),
+                "copy_recvbuf", RangePolicy<KokkosExecSpace>(0, csize),
                 KOKKOS_LAMBDA(std::size_t buffer_index) {
                     sv_view(buffer_index + offset) = recvbuf_view(buffer_index);
                 });
+            Kokkos::fence();
```
Comment (kszenes) [inline pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]: question: Why did this additional `fence()` need to be introduced?

**Review by kszenes — CHANGES_REQUESTED** [review]
Thanks, looks good. Just a couple of suggestions, please have a look.

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp (hunk 8 above)
Comment (josephleekl) [reply pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]: good call. updated.

File: pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp (hunk 15 above)
Comment (josephleekl) [reply pennylane_lightning/core/simulators/lightning_kokkos/StateVectorKokkosMPI.hpp]: it's technically not required. I'll remove it.

File: pennylane_lightning/core/simulators/lightning_kokkos/utils/MPIManagerKokkos.hpp (hunk 13 above)
Comment (josephleekl) [reply pennylane_lightning/core/simulators/lightning_kokkos/utils/MPIManagerKokkos.hpp:121]: indeed - updated the other methods too.

**Review by kszenes — APPROVED** [review]
Thanks all good!
